packages/react-native/ReactCommon/react/renderer/graphics/platform/ios/react/renderer/graphics/HostPlatformColor.h

```diff
@@ -8,32 +8,108 @@
+#include <react/utils/hash_combine.h>
 #include <cmath>
 
 namespace facebook::react {
 
-using Color = int32_t;
+enum ColorKind { UndefinedKind, ComponentsKind, DynamicKind };
+
+struct DynamicColor {
+  int32_t lightColor = 0;
+  int32_t darkColor = 0;
+  int32_t highContrastLightColor = 0;
+  int32_t highContrastDarkColor = 0;
+
+  bool operator==(const DynamicColor& other) const {
+    return lightColor == other.lightColor && darkColor == other.darkColor &&
+        highContrastLightColor == other.highContrastLightColor &&
+        highContrastDarkColor == other.highContrastDarkColor;
+  }
+};
+
+class Color {
+ public:
+  Color() : Color(UndefinedKind) {}
+  Color(DynamicColor dynamicColor)
+      : kind_(DynamicKind),
+        dynamicColor_(dynamicColor),
+        color_(dynamicColor.lightColor) {}
+  Color(ColorComponents components) {
+    float ratio = 255;
+    kind_ = ComponentsKind;
+    color_ = ((int)round(components.alpha * ratio) & 0xff) << 24 |
+        ((int)round(components.red * ratio) & 0xff) << 16 |
+        ((int)round(components.green * ratio) & 0xff) << 8 |
+        ((int)round(components.blue * ratio) & 0xff);
+  }
+  int32_t getColor() const {
+    return color_;
+  }
+  ColorKind getKind() const {
+    return kind_;
+  }
+  DynamicColor getDynamicColor() const {
+    return dynamicColor_;
+  }
+  bool operator==(const Color& other) const {
+    return kind_ == other.kind_ && color_ == other.color_ &&
+        dynamicColor_ == other.dynamicColor_;
+  }
+  bool operator!=(const Color& other) const {
+    return kind_ != other.kind_ || color_ != other.color_ ||
+        dynamicColor_ != other.dynamicColor_;
+  }
+
+ private:
+  Color(ColorKind kind) : kind_(kind), color_(0) {}
+  ColorKind kind_;
+  int32_t color_ = 0;
+  DynamicColor dynamicColor_;
+};
 
 namespace HostPlatformColor {
-static const facebook::react::Color UndefinedColor =
-    std::numeric_limits<facebook::react::Color>::max();
+static const facebook::react::Color UndefinedColor = Color();
 }
 
 inline Color hostPlatformColorFromComponents(ColorComponents components) {
-  float ratio = 255;
-  return ((int)round(components.alpha * ratio) & 0xff) << 24 |
-      ((int)round(components.red * ratio) & 0xff) << 16 |
-      ((int)round(components.green * ratio) & 0xff) << 8 |
-      ((int)round(components.blue * ratio) & 0xff);
+  return Color(components);
 }
 
 inline ColorComponents colorComponentsFromHostPlatformColor(Color color) {
   float ratio = 255;
+  int32_t primitiveColor = color.getColor();
   return ColorComponents{
-      (float)((color >> 16) & 0xff) / ratio,
-      (float)((color >> 8) & 0xff) / ratio,
-      (float)((color >> 0) & 0xff) / ratio,
-      (float)((color >> 24) & 0xff) / ratio};
+      (float)((primitiveColor >> 16) & 0xff) / ratio,
+      (float)((primitiveColor >> 8) & 0xff) / ratio,
+      (float)((primitiveColor >> 0) & 0xff) / ratio,
+      (float)((primitiveColor >> 24) & 0xff) / ratio};
 }
```

Comment:
given that `Color` is now a type, this could be moved to a function of color, so that we can do `Color.getColorComponents()`

Reply:
Seems we should keep it because Android or other platforms are still int32 type.

Reply:
I'm not following... After this change, all the platform would have access to the Color C++ class, no? 
I'm just suggesting to refactor the implementation in a method of the color object (which is available to all the platforms) so that the body of `colorComponentsFromHostPlatformColor` is just `return Color.getColorComponents()` (similarly to what has been done for `inline Color hostPlatformColorFromComponents(ColorComponents components)`.